Revise help output with default command

nafu opened the pull request:
@ggilder Thank you for maintaining great gem. 😄 

### Description
<!--- Describe your changes in detail -->

Add details information about default_command to help output.
You can see sample outputs with rspec.

#### Commander::HelpFormatter::TerminalCompact

<details>
<summary>global help</summary>

https://github.com/nafu/commander/blob/06283baedce0db38d40e7dc28af8561751648ab6/spec/help_formatters/terminal_compact_spec.rb#L12

```
  test

  something

  Commands:
        help          Display global or [command] help documentation
        install gem   Install some gem
        test          test description

  Global Options:
    -h, --help           Display help documentation
    -v, --version        Display version information
    -t, --trace          Display backtrace when an error occurs
```

</details>

<details>
<summary>global help with default_command</summary>

https://github.com/nafu/commander/blob/06283baedce0db38d40e7dc28af8561751648ab6/spec/help_formatters/terminal_compact_spec.rb#L42

```
  test

  something

  Commands: (* default)
        help          Display global or [command] help documentation
        install gem * Install some gem
        test          test description

  Global Options:
    -h, --help           Display help documentation
    -v, --version        Display version information
    -t, --trace          Display backtrace when an error occurs

  Options for install gem:
    --testing-command    Testing command
    --testing-command-second Testing command second
```

</details>

<details>
<summary>global help with default_command that has not command option</summary>

https://github.com/nafu/commander/blob/06283baedce0db38d40e7dc28af8561751648ab6/spec/help_formatters/terminal_compact_spec.rb#L81

```
  test

  something

  Commands: (* default)
        help          Display global or [command] help documentation
        install gem * Install some gem
        test          test description

  Global Options:
    -h, --help           Display help documentation
    -v, --version        Display version information
    -t, --trace          Display backtrace when an error occurs
```

</details>

#### Commander::HelpFormatter::Terminal

<details>
<summary>global help</summary>

https://github.com/nafu/commander/blob/06283baedce0db38d40e7dc28af8561751648ab6/spec/help_formatters/terminal_spec.rb#L12

```
  NAME:

    test

  DESCRIPTION:

    something

  COMMANDS:

        help          Display global or [command] help documentation
        install gem   Install some gem
        test          test description

  GLOBAL OPTIONS:

    -h, --help
        Display help documentation

    -v, --version
        Display version information

    -t, --trace
        Display backtrace when an error occurs
```

</details>

<details>
<summary>global help with default_command that has command option</summary>

https://github.com/nafu/commander/blob/06283baedce0db38d40e7dc28af8561751648ab6/spec/help_formatters/terminal_spec.rb#L41

```
  NAME:

    test

  DESCRIPTION:

    something

  COMMANDS: (* default)

        help          Display global or [command] help documentation
        install gem * Install some gem
        test          test description

  GLOBAL OPTIONS:

    -h, --help
        Display help documentation

    -v, --version
        Display version information

    -t, --trace
        Display backtrace when an error occurs

  OPTIONS for install gem:

    --testing-command
        Testing command

    --testing-command-second
        Testing command second

```

</details>

<details>
<summary>global help with default_command that has not command option</summary>

https://github.com/nafu/commander/blob/06283baedce0db38d40e7dc28af8561751648ab6/spec/help_formatters/terminal_spec.rb#L79

```
  NAME:

    test

  DESCRIPTION:

    something

  COMMANDS: (* default)

        help          Display global or [command] help documentation
        install gem * Install some gem
        test          test description

  GLOBAL OPTIONS:

    -h, --help
        Display help documentation

    -v, --version
        Display version information

    -t, --trace
        Display backtrace when an error occurs
```

</details>

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Even if you set default_command, the help command does not show the information about default_command.
It is very helpful that you can see the default_command and its options with help command. 🚀 

<!--- If it fixes an open issue, please link to the issue here. -->
<!--- Please describe in detail how you tested your changes. --->

I updated not only erb files but rspec so that you can make sure this behavior.

mfurtak commented:
This is really important for fastlane's use of Commander :+1:

hjanuschka commented:
@nafu really great - thx for this PR

ggilder commented on lib/commander/help_formatters/terminal/help.erb:
```diff
@@ -1,3 +1,6 @@
+<% unless @default_command.nil? -%>
```
Hmm, I wonder if the logic to set `default_command` here should be moved to the base help formatter, especially as it's duplicated between the two help format files.

nafu replied:
@ggilder Thank you for your review.
I'll address the feedbacks!

ggilder commented on lib/commander/help_formatters/terminal/help.erb:
```diff
@@ -6,10 +9,11 @@
     <%= Commander::HelpFormatter.indent 4, program(:description) %>
 
-  <%= $terminal.color "COMMANDS", :bold %>:
+  <%= "%s: %s" % [$terminal.color("COMMANDS", :bold), default_command.nil? ? '' : '(* default)'] %>
```
I would skip the explicit `.nil?` and the ternary here, this is effectively the same as the more concise `('(* default)' if default_command)`

ggilder replied:
Also this doesn't seem that clear to me... instead of indicating the default with an asterisk, why not simply label the actual default command below with "(default)" ? That seems more obvious.

nafu replied:
@ggilder You're right. I also think direct use of "(default)" is more clear, but * would be more readable. You can see in the real example below.

### With *

<details>
<summary>with * would be readable</summary>

```
  fastlane

  CLI for 'fastlane' - The easiest way to automate beta deployments and releases for your iOS and
Android apps

        Run using `fastlane [platform] [lane_name]`
        To pass values to the lanes use `fastlane [platform] [lane_name] key:value key2:value2`

  Commands: (* default)
        action                 Shows more information for a specific command
        actions                Lists all available fastlane actions
        add_plugin             Add a new plugin to your fastlane setup
        docs                   Generate a markdown based documentation based on the Fastfile
        enable_auto_complete   Enable tab auto completion
        env                    Print your fastlane environment, use this when you submit an issue on
GitHub
        help                   Display global or [command] help documentation
        init                   Helps you with your initial fastlane setup
        install_plugins        Install all plugins for this project
        lanes                  Lists all available lanes and shows their description
        list                   Lists all available lanes without description
        new_action             Create a new custom action for fastlane.
        new_plugin             Create a new plugin that can be used with fastlane
        run                    Run a fastlane one-off action without a full lane
        search_plugins         Search for plugins, search query is optional
        trigger              * Run a specific lane. Pass the lane name and optionally the platform
first.
        update_fastlane        Update fastlane to the latest release
        update_plugins         Update all plugin dependencies

  Global Options:
    --verbose
    --capture_output     Captures the output of the current run, and generates a markdown issue
template
    --troubleshoot       Enables extended verbose mode. Use with caution, as this even includes ALL
sensitive data. Cannot be used on CI.
    --env STRING[,STRING2] Add environment(s) to use with `dotenv`
    -h, --help           Display help documentation
    -v, --version        Display version information

  Options for trigger:
    --env STRING[,STRING2] Add environment(s) to use with `dotenv`

```

</details>

### With (default)

<details>
<summary>with (default) to indicate default_command directly</summary>

```
  fastlane

  CLI for 'fastlane' - The easiest way to automate beta deployments and releases for your iOS and
Android apps

        Run using `fastlane [platform] [lane_name]`
        To pass values to the lanes use `fastlane [platform] [lane_name] key:value key2:value2`

  Commands:
        action               Shows more information for a specific command
        actions              Lists all available fastlane actions
        add_plugin           Add a new plugin to your fastlane setup
        docs                 Generate a markdown based documentation based on the Fastfile
        enable_auto_complete Enable tab auto completion
        env                  Print your fastlane environment, use this when you submit an issue on
GitHub
        help                 Display global or [command] help documentation
        init                 Helps you with your initial fastlane setup
        install_plugins      Install all plugins for this project
        lanes                Lists all available lanes and shows their description
        list                 Lists all available lanes without description
        new_action           Create a new custom action for fastlane.
        new_plugin           Create a new plugin that can be used with fastlane
        run                  Run a fastlane one-off action without a full lane
        search_plugins       Search for plugins, search query is optional
        trigger (default)    Run a specific lane. Pass the lane name and optionally the platform
first.
        update_fastlane      Update fastlane to the latest release
        update_plugins       Update all plugin dependencies

  Global Options:
    --verbose
    --capture_output     Captures the output of the current run, and generates a markdown issue
template
    --troubleshoot       Enables extended verbose mode. Use with caution, as this even includes ALL
sensitive data. Cannot be used on CI.
    --env STRING[,STRING2] Add environment(s) to use with `dotenv`
    -h, --help           Display help documentation
    -v, --version        Display version information

  Options for trigger:
    --env STRING[,STRING2] Add environment(s) to use with `dotenv`

```

</details>

<details>
<summary>with (default) after command can be too long</summary>

```
  Commands:
        action                          Shows more information for a specific command
        actions                         Lists all available fastlane actions
        add_plugin                      Add a new plugin to your fastlane setup
        docs                            Generate a markdown based documentation based on the Fastfile
        enable_auto_complete (default)  Enable tab auto completion
```

</details>

ggilder commented on lib/commander/help_formatters/terminal/help.erb:
```diff
@@ -6,10 +9,11 @@
 <% for name, command in @commands.sort -%>
 	<% unless alias? name %>
-    <%= "%-#{max_command_length}s %s" % [command.name, command.summary || command.description] -%>
+    <% summary_prefix = (!default_command.nil? && command.name == default_command.name)? '* ' : '  ' -%>
```
We may be missing a spec for this, but does this increase the indentation level of existing help output in cases where there's _not_ a default command? It would be nice to avoid changing the current output style.

nafu replied:
Yes, I'll fix that.

KrauseFx commented:
Thanks for the review, so glad to see a super fast response time here @ggilder 👍 Thanks for your help

nafu commented:
@ggilder I addressed the feedback, and left the comment about [mark for default_command](https://github.com/commander-rb/commander/pull/44#discussion_r105842176). Thanks for your help.

#### Commander::HelpFormatter::TerminalCompact

<details>
<summary>global help</summary>

https://github.com/nafu/commander/blob/06283baedce0db38d40e7dc28af8561751648ab6/spec/help_formatters/terminal_compact_spec.rb#L12

```
  test

  something

  Commands:
    help        Display global or [command] help documentation
    install gem Install some gem
    test        test description

  Global Options:
    -h, --help           Display help documentation
    -v, --version        Display version information
    -t, --trace          Display backtrace when an error occurs
```

</details>

<details>
<summary>global help with default_command</summary>

https://github.com/nafu/commander/blob/06283baedce0db38d40e7dc28af8561751648ab6/spec/help_formatters/terminal_compact_spec.rb#L42

```
  test

  something

  Commands: (* default)
    help          Display global or [command] help documentation
    install gem * Install some gem
    test          test description

  Global Options:
    -h, --help           Display help documentation
    -v, --version        Display version information
    -t, --trace          Display backtrace when an error occurs

  Options for install gem:
    --testing-command    Testing command
    --testing-command-second Testing command second
```

</details>

<details>
<summary>global help with default_command that has not command option</summary>

https://github.com/nafu/commander/blob/06283baedce0db38d40e7dc28af8561751648ab6/spec/help_formatters/terminal_compact_spec.rb#L81

```
  test

  something

  Commands: (* default)
    help          Display global or [command] help documentation
    install gem * Install some gem
    test          test description

  Global Options:
    -h, --help           Display help documentation
    -v, --version        Display version information
    -t, --trace          Display backtrace when an error occurs
```

</details>

#### Commander::HelpFormatter::Terminal

<details>
<summary>global help</summary>

https://github.com/nafu/commander/blob/06283baedce0db38d40e7dc28af8561751648ab6/spec/help_formatters/terminal_spec.rb#L12

```
  NAME:

    test

  DESCRIPTION:

    something

  COMMANDS:

    help        Display global or [command] help documentation
    install gem Install some gem
    test        test description

  GLOBAL OPTIONS:

    -h, --help
        Display help documentation

    -v, --version
        Display version information

    -t, --trace
        Display backtrace when an error occurs
```

</details>

<details>
<summary>global help with default_command that has command option</summary>

https://github.com/nafu/commander/blob/06283baedce0db38d40e7dc28af8561751648ab6/spec/help_formatters/terminal_spec.rb#L41

```
  NAME:

    test

  DESCRIPTION:

    something

  COMMANDS: (* default)

    help          Display global or [command] help documentation
    install gem * Install some gem
    test          test description

  GLOBAL OPTIONS:

    -h, --help
        Display help documentation

    -v, --version
        Display version information

    -t, --trace
        Display backtrace when an error occurs

  OPTIONS for install gem:

    --testing-command
        Testing command

    --testing-command-second
        Testing command second
```

</details>

<details>
<summary>global help with default_command that has not command option</summary>

https://github.com/nafu/commander/blob/06283baedce0db38d40e7dc28af8561751648ab6/spec/help_formatters/terminal_spec.rb#L79

```
  NAME:

    test

  DESCRIPTION:

    something

  COMMANDS: (* default)

    help          Display global or [command] help documentation
    install gem * Install some gem
    test          test description

  GLOBAL OPTIONS:

    -h, --help
        Display help documentation

    -v, --version
        Display version information

    -t, --trace
        Display backtrace when an error occurs
```

</details>

nafu commented:
@ggilder I also added another specs to make sure default_command behavior, could you review again? Thanks!

KrauseFx commented:
Hey @ggilder, let us know if there is anything else you need from our side. Once that change is merged, we'd love a new release for this, so we can start using it across all of @fastlane. No worries if it takes longer, just let us know so that we can use a fork in the mean-time 👍

ggilder commented:
Hi, sorry for the delay reviewing, been busy with the day job :)

I think the major snag for me is around the formatting. The asterisk feels kind of clunky, and I'm wondering if just adding another section would make more sense. Example:

```
NAME:

    test

  DESCRIPTION:

    something

  COMMANDS:

    help          Display global or [command] help documentation
    install gem   Install some gem
    test          test description

  DEFAULT COMMAND:

    install gem

  GLOBAL OPTIONS:
....
```

Honestly it's hard because a "default command" is kind of a weird concept. Most of the oft-cited examples of "monolith" CLIs like git don't have them, so there's not much prior art as far as how a default command would be indicated in help text. Please direct me to examples if you know of any, I'm curious to see how this handled elsewhere!

ggilder commented:
Closing since this has not been updated in a while. Please feel free to reopen if this is still an issue for you.

KrauseFx commented:
Thanks, we're using a fork for fastlane, and everything's working great there 👍

ggilder commented:
Ok. You may want to rebase on top of the changes from 4.4.4 which I just released.

nafu commented:
Thanks :+1:
